st2actions/st2actions/runners/ssh/paramiko_ssh.py

```diff
@@ -0,0 +1,406 @@
+            strip = os.path.dirname(local_path)
+        else:
+            strip = os.path.dirname(os.path.dirname(local_path))
+
+        remote_paths = []
+
+        for context, dirs, files in os.walk(local_path):
+            rcontext = context.replace(strip, '', 1)
+            # normalize pathname separators with POSIX separator
+            rcontext = rcontext.replace(os.sep, '/')
+            rcontext = rcontext.lstrip('/')
+            rcontext = posixpath.join(remote_path, rcontext)
+
+            if not self.exists(rcontext):
+                self.sftp.mkdir(rcontext)
+
+            for d in dirs:
+                n = posixpath.join(rcontext, d)
+                if not self.exists(n):
+                    self.sftp.mkdir(n)
+
+            for f in files:
+                local_path = os.path.join(context, f)
+                n = posixpath.join(rcontext, f)
+                p = self.put(local_path=local_path, remote_path=n,
+                             mirror_local_mode=mirror_local_mode, mode=mode)
+                remote_paths.append(p)
+
+        return remote_paths
+
+    def create_file(self, path, contents=None, chmod=None, mode='w'):
+        extra = {'_path': path, '_mode': mode, '_chmod': chmod}
+        self.logger.debug('Uploading file', extra=extra)
+
+        # less than ideal, but we need to mkdir stuff otherwise file() fails
+        head, tail = psplit(path)
+
+        if path[0] == "/":
+            self.sftp.chdir("/")
+        else:
+            # Relative path - start from a home directory (~)
+            self.sftp.chdir('.')
+
+        for part in head.split("/"):
+            if part != "":
+                try:
+                    self.sftp.mkdir(part)
+                except IOError:
+                    # so, there doesn't seem to be a way to
+                    # catch EEXIST consistently *sigh*
+                    pass
+                self.sftp.chdir(part)
+
+        cwd = self.sftp.getcwd()
+
+        ak = self.sftp.file(tail, mode=mode)
+        ak.write(contents)
+        if chmod is not None:
+            ak.chmod(chmod)
+        ak.close()
+
+        if path[0] == '/':
+            file_path = path
+        else:
+            file_path = pjoin(cwd, path)
+
+        return file_path
+
+    def exists(self, remote_path):
+        try:
+            self.sftp.lstat(remote_path).st_mode
+        except IOError:
+            return False
+
+        return True
+
+    def mkdir(self, dir_path):
+        extra = {'_dir_path': dir_path}
+        self.logger.debug('mkdir', extra=extra)
+        self.sftp.mkdir(dir_path)
+        return True
+
+    def delete_file(self, path):
+        extra = {'_path': path}
+        self.logger.debug('Deleting file', extra=extra)
+        self.sftp.unlink(path)
+        return True
+
+    def delete_dir(self, path, force=False, timeout=None):
+        extra = {'_path': path}
+        if force:
+            command = 'rm -rf %s' % path
```

Comment:
Ok, I took a detailed look and it appears you have to shlex everything before giving it to fabric. Since there are several layers before the actual paramiko works with these variables and shlexing can happen anywhere, the whole path seems fragile. We could probably look into making quote_unix more correct.